routing/fake_graph.hpp

```diff
@@ -0,0 +1,146 @@
+
+    for (auto const & kv : rhs.m_ingoing)
+      m_ingoing[kv.first].insert(kv.second.begin(), kv.second.end());
+
+    for (auto const & kv : rhs.m_realToFake)
+      m_realToFake[kv.first].insert(kv.second.begin(), kv.second.end());
+
+    m_fakeToReal.insert(rhs.m_fakeToReal.begin(), rhs.m_fakeToReal.end());
+  }
+
+  // Returns Vertex which corresponds |segment|. Segment must be a part of the fake graph.
+  VertexType const & GetVertex(SegmentType const & segment) const
+  {
+    auto const it = m_segmentToVertex.find(segment);
+    CHECK(it != m_segmentToVertex.end(), ("Vertex for invalid fake segment requested."));
+    return it->second;
+  }
+
+  // Returns outgoing/ingoing edges set for specified segment.
+  std::set<SegmentType> const & GetEdges(SegmentType const & segment, bool isOutgoing) const
+  {
+    auto const & adjacentEdges = isOutgoing ? m_outgoing : m_ingoing;
+
+    auto const it = adjacentEdges.find(segment);
+    if (it != adjacentEdges.end())
+      return it->second;
+
+    return kEmptySet;
+  }
+
+  size_t GetSize() const { return m_segmentToVertex.size(); }
+
+  std::set<SegmentType> const & GetFake(RealType const & real) const
+  {
+    auto const it = m_realToFake.find(real);
+    if (it != m_realToFake.end())
+      return it->second;
+
+    return kEmptySet;
+  }
+
+  bool FindReal(SegmentType const & fake, RealType & real) const
+  {
+    auto const it = m_fakeToReal.find(fake);
+    if (it == m_fakeToReal.end())
+      return false;
+
+    real = it->second;
+    return true;
+  }
+
+  bool FindSegment(VertexType const & vertex, SegmentType & segment) const
+  {
+    auto const it = m_vertexToSegment.find(vertex);
+    if (it == m_vertexToSegment.end())
+      return false;
+
+    segment = it->second;
+    return true;
+  }
+
+  std::map<SegmentType, RealType> const & GetFakeToReal() const { return m_fakeToReal; }
+
+private:
+  // Key is fake segment, value is set of outgoing fake segments.
+  std::map<SegmentType, std::set<SegmentType>> m_outgoing;
+  // Key is fake segment, value is set of ingoing fake segments.
+  std::map<SegmentType, std::set<SegmentType>> m_ingoing;
+  // Key is fake segment, value is fake vertex which corresponds fake segment.
+  std::map<SegmentType, VertexType> m_segmentToVertex;
+  // Key is fake vertex value is fake segment which corresponds fake vertex.
+  std::map<VertexType, SegmentType> m_vertexToSegment;
+  // Key is fake segment of type VertexType::Type::PartOfReal, value is corresponding real segment.
+  std::map<SegmentType, RealType> m_fakeToReal;
+  // Key is real segment, value is set of fake segments with type VertexType::Type::PartOfReal
+  // which are parts of this real segment.
+  std::map<RealType, std::set<SegmentType>> m_realToFake;
+  // To return empty set by const reference.
+  static const std::set<SegmentType> kEmptySet;
+};
+
+template <typename SegmentType, typename VertexType, typename RealType>
```

Comment:
I really dislike the idea to incorporate this static into hpp, because it will be duplicated in every translation unit that uses the hpp. Why not to just drop the "static" keyword?

Reply:
да, так лучше, хоть по смыслу он и статический, сделала